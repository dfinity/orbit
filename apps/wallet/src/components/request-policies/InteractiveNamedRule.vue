<template>
  <template v-if="loading">
    <VProgressCircular indeterminate />
  </template>
  <template v-else-if="error !== null"> {{ id }} <ErrorTooltip :text="error" /> </template>
  <template v-else-if="namedRule">
    <VTooltip v-if="tooltip" location="bottom" content-class="white-space-pre-wrap" :text="tooltip">
      <template #activator="{ props: activatorProps }">
        <span v-bind="activatorProps" class="underline-dotted font-weight-bold">
          {{ namedRule.name }}
        </span>
      </template>
    </VTooltip>
  </template>
  <template v-else>
    {{ id }}
  </template>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { onMounted } from 'vue';
import { computed } from 'vue';
import { toRefs } from 'vue';
import ErrorTooltip from '~/components/error/ErrorTooltip.vue';
import { useRuleToTooltip } from '~/composables/request-policies.composable';
import { NamedRule, UUID } from '~/generated/station/station.did';
import { services } from '~/plugins/services.plugin';
import { useAppStore } from '~/stores/app.store';
import { getErrorMessage } from '~/utils/error.utils';

const input = defineProps<{
  id: UUID;
}>();

const props = toRefs(input);
const namedRule = ref<NamedRule | null>(null);
const loading = ref(true);
const error = ref<string | null>(null);
const appStore = useAppStore();

const rule = computed(() => (namedRule.value ? namedRule.value.rule : null));
const tooltip = useRuleToTooltip(rule);

onMounted(async () => {
  loading.value = true;
  try {
    namedRule.value = (await services().station.getNamedRule(props.id.value)).named_rule;
  } catch (e) {
    appStore.sendErrorNotification(e);
    error.value = getErrorMessage(e);
  } finally {
    loading.value = false;
  }
});
</script>
