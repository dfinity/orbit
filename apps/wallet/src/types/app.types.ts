import en from '~/locales/en.locale';
import { BreadCrumbItem } from '~/types/navigation.types';

export interface AppInitConfig {
  name: string;
  version: string;
  logLevel: 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'silent';
  baseUrl: string;
  versionedBaseUrl: string;
  isProduction: boolean;
  buildMode: string;
  apiGatewayUrl: URL;
  locale: {
    default: string;
    supportedLocales: string[];
  };
  providers: {
    internetIdentity: string;
  };
  canisters: {
    app_wallet: string;
    controlPanel: string;
    internetIdentity: string;
    icpIndex: string;
  };
}

export enum SupportedTheme {
  Dark = 'dark',
  Light = 'light',
}

export type AppTranslations = typeof en;

export interface GlobalNotification {
  show: boolean;
  message: string;
  type: 'error' | 'success' | 'info' | 'warning';
}

export interface Pagination {
  limit: number;
  totalPages: number;
  selectedPage: number;
}

export interface TableHeaderProps {
  class?: string;
}

export interface TableHeader {
  title: string;
  key: string;
  sortable?: boolean;
  headerProps?: TableHeaderProps;
}

export interface PageProps {
  title?: string;
  subtitle?: string;
  breadcrumbs?: BreadCrumbItem[];
}

export type CsvRow = Record<string, string>;

export interface CsvTable {
  headers: CsvRow;
  rows: CsvRow[];
}

export enum CyclesUnit {
  Trillion = 'TC',
  Billion = 'BC',
  Million = 'MC',
  // Can be used to display cycles in the smallest unit.
  Smallest = 'e8s',
}
