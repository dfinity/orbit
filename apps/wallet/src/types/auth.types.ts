export enum Privilege {
  Capabilities = 'Capabilities',
  SystemInfo = 'SystemInfo',
  AddUserGroup = 'AddUserGroup',
  ListUserGroups = 'ListUserGroups',
  AddUser = 'AddUser',
  ListUsers = 'ListUsers',
  AddRequestPolicy = 'AddRequestPolicy',
  ListRequestPolicies = 'ListRequestPolicies',
  ListAccounts = 'ListAccounts',
  ListPermissions = 'ListPermissions',
  AddAccount = 'AddAccount',
  ListAddressBookEntries = 'ListAddressBookEntries',
  AddAddressBookEntry = 'AddAddressBookEntry',
  ListRequests = 'ListRequests',
  SystemUpgrade = 'SystemUpgrade',
  ManageSystemInfo = 'ManageSystemInfo',
  ListExternalCanisters = 'ListExternalCanisters',
  CreateExternalCanister = 'CreateExternalCanister',
  CallAnyExternalCanister = 'CallAnyExternalCanister',
}

export enum RequiredSessionState {
  Authenticated = 'authenticated',
  AuthenticatedNoStation = 'authenticated-no-station',
  AuthenticatedHasStations = 'authenticated-has-stations',
  ConnectedToStation = 'connected-to-station',
  Guest = 'guest',
  Any = 'any',
}

export interface PermissionRequirements {
  session: RequiredSessionState;
  privileges?: Privilege[];
}

export interface AuthRouteMeta {
  check: PermissionRequirements;
}

declare module 'vue-router' {
  interface RouteMeta {
    // must be declared by every route
    auth: AuthRouteMeta;
  }
}
