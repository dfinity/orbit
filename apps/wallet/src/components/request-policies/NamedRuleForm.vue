<template>
  <VForm ref="form" @submit.prevent="submit">
    <VTextField
      v-if="model.id && props.display.value.id"
      v-model="model.id"
      name="id"
      :label="$t('terms.id')"
      variant="plain"
      density="comfortable"
      disabled
      :prepend-icon="mdiIdentifier"
    />

    <VTextField
      v-model="model.name"
      name="name"
      :label="$t('terms.name')"
      density="comfortable"
      :rules="[requiredRule]"
      :prepend-icon="mdiFileDocumentCheckOutline"
      :disabled="isViewMode"
    />

    <VTextField
      v-model="description"
      name="description"
      :label="$t('terms.description')"
      density="comfortable"
      :prepend-icon="mdiInformationBoxOutline"
      :disabled="isViewMode"
    />

    <div class="mt-4 mb-2 text-body-1">{{ $t('terms.rule') }}</div>
    <RuleBuilder v-model="model.rule" :disabled="isViewMode" @remove="model.rule = undefined" />

    <span v-if="!model.rule && isViewMode">
      {{ $t('terms.none') }}
    </span>
  </VForm>
</template>

<script lang="ts" setup>
import { mdiFileDocumentCheckOutline, mdiIdentifier, mdiInformationBoxOutline } from '@mdi/js';
import { computed, ref, toRefs, watch } from 'vue';
import RuleBuilder from '~/components/request-policies/rule/RuleBuilder.vue';
import { NamedRule } from '~/generated/station/station.did';
import { VFormValidation } from '~/types/helper.types';
import { requiredRule } from '~/utils/form.utils';

export type RequestPolicyFormProps = {
  modelValue: Partial<NamedRule>;
  valid?: boolean;
  mode?: 'view' | 'edit';
  display?: {
    id?: boolean;
  };
};

const form = ref<VFormValidation | null>(null);

const input = withDefaults(defineProps<RequestPolicyFormProps>(), {
  valid: true,
  display: () => ({
    id: true,
  }),
  mode: 'edit',
});
const props = toRefs(input);

const emit = defineEmits<{
  (event: 'update:modelValue', payload: RequestPolicyFormProps['modelValue']): void;
  (event: 'valid', payload: boolean): void;
  (event: 'submit', payload: RequestPolicyFormProps['modelValue']): void;
}>();

const model = computed(() => props.modelValue.value);

watch(model.value, newValue => emit('update:modelValue', newValue), { deep: true });

const description = computed({
  get: () => (model.value.description ? model.value.description[0] : ''),
  set: value => {
    model.value.description = value ? [value] : undefined;
  },
});

const isViewMode = computed(() => props.mode.value === 'view');

const isFormValid = computed(() => (form.value ? form.value.isValid : false));
watch(
  () => isFormValid.value,
  isValid => emit('valid', isValid ?? false),
);

const submit = async () => {
  const { valid } = form.value ? await form.value.validate() : { valid: false };

  if (valid) {
    emit('submit', model.value);
  }
};
</script>
