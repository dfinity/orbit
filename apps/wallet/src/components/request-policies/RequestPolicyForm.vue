<template>
  <VForm ref="form" @submit.prevent="submit">
    <VTextField
      v-if="model.id && props.display.value.id"
      v-model="model.id"
      name="id"
      :label="$t('terms.id')"
      variant="plain"
      density="compact"
      disabled
    />

    <SpecifierSelector
      v-model="model.specifier"
      :disabled="isViewMode"
      @changed-variant="onChangedVariant"
    />

    <template v-if="model.specifier">
      <div class="mt-4 mb-2 text-body-2">{{ $t('terms.rule') }}</div>
      <RuleBuilder
        v-model="model.rule"
        :specifier="model.specifier"
        :disabled="isViewMode"
        @remove="model.rule = undefined"
      />

      <span v-if="!model.rule && isViewMode">
        {{ $t('terms.none') }}
      </span>
    </template>
  </VForm>
</template>

<script lang="ts" setup>
import { computed, ref, toRefs } from 'vue';
import RuleBuilder from '~/components/request-policies/rule/RuleBuilder.vue';
import SpecifierSelector from '~/components/request-policies/specifier/SpecifierSelector.vue';
import { RequestPolicy } from '~/generated/station/station.did';
import { VFormValidation } from '~/types/helper.types';

export type RequestPolicyFormProps = {
  modelValue: Partial<RequestPolicy>;
  valid?: boolean;
  mode?: 'view' | 'edit';
  display?: {
    id?: boolean;
    specifier?: boolean;
  };
};

const form = ref<VFormValidation | null>(null);

const input = withDefaults(defineProps<RequestPolicyFormProps>(), {
  valid: true,
  display: () => ({
    id: true,
    specifier: true,
  }),
  mode: 'edit',
});
const props = toRefs(input);

const emit = defineEmits<{
  (event: 'update:modelValue', payload: RequestPolicyFormProps['modelValue']): void;
  (event: 'valid', payload: boolean): void;
  (event: 'submit', payload: RequestPolicyFormProps['modelValue']): void;
}>();

const model = computed({
  get: () => props.modelValue.value,
  set: value => emit('update:modelValue', value),
});

const isViewMode = computed(() => props.mode.value === 'view');

const onChangedVariant = (): void => {
  model.value.rule = undefined;
};

const submit = async () => {
  const { valid } = form.value ? await form.value.validate() : { valid: false };

  if (valid) {
    emit('submit', model.value);
  }
};
</script>
