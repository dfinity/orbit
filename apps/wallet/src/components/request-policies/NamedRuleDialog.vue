<template>
  <VDialog
    v-model="openModel"
    :persistent="loading || saving"
    transition="dialog-bottom-transition"
    scrollable
    :max-width="props.dialogMaxWidth.value"
  >
    <DataLoader
      v-slot="{ data }"
      :load="loadNamedRule"
      @loading="loading = $event"
      @loaded="namedRule = $event"
    >
      <VCard>
        <VToolbar color="background">
          <VToolbarTitle>{{ $t('pages.approval_rules.dialog.title') }}</VToolbarTitle>
          <VBtn :disabled="loading || saving" :icon="mdiClose" @click="openModel = false" />
        </VToolbar>
        <VCardText v-if="loading" class="py-8">
          <LoadingMessage />
        </VCardText>
        <VCardText v-else>
          <NamedRuleForm
            v-if="data"
            v-model="namedRule"
            :mode="props.readonly.value ? 'view' : 'edit'"
            @submit="save"
            @valid="valid = $event"
          />
        </VCardText>
        <VDivider />
        <VCardActions class="pa-3">
          <VSpacer />
          <VBtn
            v-if="!props.readonly.value"
            color="primary"
            variant="elevated"
            :disabled="!canSave"
            :loading="saving"
            data-test-id="save-named-rule"
            @click="save"
          >
            {{ $t('terms.save') }}
          </VBtn>
        </VCardActions>
      </VCard>
    </DataLoader>
  </VDialog>
</template>
<script lang="ts" setup>
import { mdiClose } from '@mdi/js';
import { computed, Ref, ref, toRefs } from 'vue';
import {
  VBtn,
  VCard,
  VCardActions,
  VCardText,
  VDialog,
  VDivider,
  VSpacer,
  VToolbar,
  VToolbarTitle,
} from 'vuetify/components';
import DataLoader from '~/components/DataLoader.vue';
import LoadingMessage from '~/components/LoadingMessage.vue';
import {
  useOnFailedOperation,
  useOnSuccessfulOperation,
} from '~/composables/notifications.composable';
import logger from '~/core/logger.core';
import { NamedRule, UUID } from '~/generated/station/station.did';
import { useStationStore } from '~/stores/station.store';
import { assertAndReturn } from '~/utils/helper.utils';
import NamedRuleForm from './NamedRuleForm.vue';

const input = withDefaults(
  defineProps<{
    namedRuleId?: UUID;
    open?: boolean;
    dialogMaxWidth?: number;
    readonly?: boolean;
  }>(),
  {
    namedRuleId: undefined,
    open: false,
    dialogMaxWidth: 800,
    readonly: false,
  },
);

const emit = defineEmits<{
  (event: 'update:open', payload: boolean): void;
}>();

const props = toRefs(input);
const valid = ref(false);
const loading = ref(false);
const saving = ref(false);
const namedRule: Ref<Partial<NamedRule>> = ref({});
const openModel = computed({
  get: () => props.open.value,
  set: value => emit('update:open', value),
});

const station = useStationStore();

const loadNamedRule = async (): Promise<Partial<NamedRule>> => {
  if (props.namedRuleId.value === undefined) {
    const createModel: Partial<NamedRule> = {
      rule: undefined,
      name: '',
      description: [],
    };

    return createModel;
  }
  const result = await station.service.getNamedRule(props.namedRuleId.value);
  return result.named_rule;
};

const canSave = computed(() => {
  return valid.value && !loading.value && !!namedRule.value?.rule;
});

const save = async (): Promise<void> => {
  if (!canSave.value) {
    return;
  }

  try {
    saving.value = true;
    if (namedRule.value.id) {
      const request = await station.service.editNamedRule({
        named_rule_id: namedRule.value.id,
        name: [assertAndReturn(namedRule.value.name)],
        description: namedRule.value.description ? [namedRule.value.description] : [['']],
        rule: [assertAndReturn(namedRule.value.rule)],
      });

      useOnSuccessfulOperation(request);

      openModel.value = false;
      return;
    }

    const request = await station.service.addNamedRule({
      name: assertAndReturn(namedRule.value.name),
      description: assertAndReturn(namedRule.value.description),
      rule: assertAndReturn(namedRule.value.rule),
    });

    useOnSuccessfulOperation(request);

    openModel.value = false;
  } catch (error) {
    logger.error(`Failed to save named rule ${error}`);

    useOnFailedOperation();
  } finally {
    saving.value = false;
  }
};
</script>
