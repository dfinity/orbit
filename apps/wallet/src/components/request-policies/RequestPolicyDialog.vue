<template>
  <VDialog
    v-model="openModel"
    :persistent="loading || saving"
    transition="dialog-bottom-transition"
    scrollable
    :max-width="props.dialogMaxWidth.value"
  >
    <DataLoader
      v-slot="{ data }"
      :load="loadPolicy"
      @loading="loading = $event"
      @loaded="requestPolicy = $event.policy"
    >
      <VCard>
        <VToolbar color="background">
          <VToolbarTitle>{{ $t('pages.request_policies.dialog_title') }}</VToolbarTitle>
          <VBtn :disabled="loading || saving" :icon="mdiClose" @click="openModel = false" />
        </VToolbar>
        <VCardText v-if="loading" class="py-8">
          <LoadingMessage />
        </VCardText>
        <VCardText v-else>
          <RequestPolicyForm
            v-if="data"
            v-model="requestPolicy"
            :mode="props.readonly.value ? 'view' : 'edit'"
            @submit="save"
            @valid="valid = $event"
          />
        </VCardText>
        <VDivider />
        <VCardActions class="pa-3">
          <VSpacer />
          <VBtn
            v-if="!props.readonly.value"
            color="primary"
            variant="elevated"
            :disabled="!canSave"
            :loading="saving"
            @click="save"
          >
            {{ $t('terms.save') }}
          </VBtn>
        </VCardActions>
      </VCard>
    </DataLoader>
  </VDialog>
</template>
<script lang="ts" setup>
import { mdiClose } from '@mdi/js';
import { computed, ref, toRefs } from 'vue';
import {
  VBtn,
  VCard,
  VCardActions,
  VCardText,
  VDialog,
  VDivider,
  VSpacer,
  VToolbar,
  VToolbarTitle,
} from 'vuetify/components';
import DataLoader from '~/components/DataLoader.vue';
import LoadingMessage from '~/components/LoadingMessage.vue';
import RequestPolicyForm from '~/components/request-policies/RequestPolicyForm.vue';
import {
  useOnFailedOperation,
  useOnSuccessfulOperation,
} from '~/composables/notifications.composable';
import logger from '~/core/logger.core';
import { RequestPolicy, UUID } from '~/generated/station/station.did';
import { useStationStore } from '~/stores/station.store';
import { assertAndReturn } from '~/utils/helper.utils';

const input = withDefaults(
  defineProps<{
    policyId?: UUID;
    open?: boolean;
    dialogMaxWidth?: number;
    readonly?: boolean;
  }>(),
  {
    policyId: undefined,
    open: false,
    dialogMaxWidth: 800,
    readonly: false,
  },
);

const emit = defineEmits<{
  (event: 'update:open', payload: boolean): void;
}>();

const props = toRefs(input);
const valid = ref(true);
const loading = ref(false);
const saving = ref(false);
const requestPolicy = ref<Partial<RequestPolicy>>({});
const openModel = computed({
  get: () => props.open.value,
  set: value => emit('update:open', value),
});

const station = useStationStore();

const loadPolicy = async (): Promise<{
  policy: Partial<RequestPolicy>;
}> => {
  if (props.policyId.value === undefined) {
    const createModel: Partial<RequestPolicy> = {
      rule: { AutoApproved: null },
    };

    return { policy: createModel };
  }

  const result = await station.service.getRequestPolicy(props.policyId.value, true);
  return result;
};

const canSave = computed(() => {
  return (
    valid.value && !loading.value && !!requestPolicy.value?.rule && !!requestPolicy.value?.specifier
  );
});

const save = async (): Promise<void> => {
  if (!canSave.value) {
    return;
  }

  try {
    saving.value = true;
    if (requestPolicy.value.id) {
      const request = await station.service.editRequestPolicy({
        policy_id: requestPolicy.value.id,
        specifier: [assertAndReturn(requestPolicy.value.specifier)],
        rule: [assertAndReturn(requestPolicy.value.rule)],
      });

      useOnSuccessfulOperation(request);

      openModel.value = false;
      return;
    }

    const request = await station.service.addRequestPolicy({
      specifier: assertAndReturn(requestPolicy.value.specifier),
      rule: assertAndReturn(requestPolicy.value.rule),
    });

    useOnSuccessfulOperation(request);

    openModel.value = false;
  } catch (error) {
    logger.error(`Failed to save request policy ${error}`);

    useOnFailedOperation();
  } finally {
    saving.value = false;
  }
};
</script>
