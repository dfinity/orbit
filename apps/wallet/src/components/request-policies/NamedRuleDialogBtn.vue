<template>
  <VBtn
    :size="props.size.value"
    :variant="props.variant.value"
    :icon="props.icon.value && !props.text.value"
    :color="props.color.value"
    @click="open = true"
  >
    <VIcon v-if="props.icon.value" :icon="props.icon.value" />
    <span v-if="props.text">{{ props.text.value }}</span>
  </VBtn>

  <NamedRuleDialog
    :open="open"
    :named-rule-id="props.namedRuleId.value"
    :readonly="props.readonly.value"
    @update:open="
      openEvent => {
        open = openEvent;

        emit('opened', openEvent);
      }
    "
  />
</template>
<script lang="ts" setup>
import { ref, toRefs } from 'vue';
import { VBtn, VIcon } from 'vuetify/components';
import { UUID } from '~/generated/station/station.did';
import NamedRuleDialog from './NamedRuleDialog.vue';

const input = withDefaults(
  defineProps<{
    namedRuleId?: UUID;
    icon?: string;
    text?: string;
    size?: 'x-small' | 'small' | 'default' | 'medium' | 'large' | 'x-large';
    variant?: 'flat' | 'text' | 'outlined' | 'elevated';
    color?: string;
    readonly?: boolean;
  }>(),
  {
    namedRuleId: undefined,
    icon: undefined,
    text: undefined,
    size: 'default',
    variant: 'elevated',
    color: 'primary',
    readonly: false,
  },
);

const open = ref(false);

const emit = defineEmits<{
  (event: 'opened', payload: boolean): void;
}>();

const props = toRefs(input);
</script>
