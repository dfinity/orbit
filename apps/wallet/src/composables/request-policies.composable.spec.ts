import { describe, expect, it } from 'vitest';
import { PopulatedRule, populatedRuleToTooltip } from './request-policies.composable';
import { RequestPolicyRuleEnum, RequestPolicyRuleUserSpecifierEnum } from '~/types/station.types';
import { i18n } from '~/plugins/i18n.plugin';

describe('request-policies.composable', () => {
  it('display the any user quorum rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.Quorum,
      approvers: {
        kind: RequestPolicyRuleUserSpecifierEnum.Any,
      },
      n: 5,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });
  it('display the some users quorum rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.Quorum,
      approvers: {
        kind: RequestPolicyRuleUserSpecifierEnum.Id,
        users: [
          {
            id: '1',
            name: 'Alice',
          },
          {
            id: '2',
            name: 'Bob',
          },
        ],
      },
      n: 5,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the some groups quorum rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.Quorum,
      approvers: {
        kind: RequestPolicyRuleUserSpecifierEnum.Group,
        groups: [
          {
            id: '1',
            name: 'Admin',
          },
          {
            id: '2',
            name: 'Developers',
          },
        ],
      },
      n: 1,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the any user percentage quorum rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.QuorumPercentage,
      approvers: {
        kind: RequestPolicyRuleUserSpecifierEnum.Any,
      },
      n: 50,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the some users percentage quorum rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.QuorumPercentage,
      approvers: {
        kind: RequestPolicyRuleUserSpecifierEnum.Id,
        users: [{ id: '1', name: 'Alice' }],
      },
      n: 50,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the allowlisted by metadata rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.AllowListedByMetadata,
      key: 'test',
      value: 'test',
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the allowlisted rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.AllowListed,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the auto approved rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.AutoApproved,
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });

  it('display the all of rule correctly', () => {
    const rule: PopulatedRule = {
      kind: RequestPolicyRuleEnum.AllOf,
      rules: [
        {
          kind: RequestPolicyRuleEnum.AnyOf,
          rules: [
            {
              kind: RequestPolicyRuleEnum.AllowListed,
            },
            {
              kind: RequestPolicyRuleEnum.AutoApproved,
            },
          ],
        },
        {
          kind: RequestPolicyRuleEnum.Not,
          rule: {
            kind: RequestPolicyRuleEnum.AllowListed,
          },
        },
        {
          kind: RequestPolicyRuleEnum.Quorum,
          approvers: { kind: RequestPolicyRuleUserSpecifierEnum.Any },
          n: 1,
        },
      ],
    };
    const tooltip = populatedRuleToTooltip(rule, i18n.global);
    expect(tooltip).toMatchSnapshot();
  });
});
