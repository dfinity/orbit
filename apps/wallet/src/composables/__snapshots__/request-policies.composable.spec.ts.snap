// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`request-policies.composable > display the all of rule correctly 1`] = `
"All of the following:
    - Any of the following:
        - Address in address book
        - No approval required
    - Approved, except when:
        - Address in address book
    - 1 approval from any user"
`;

exports[`request-policies.composable > display the allowlisted by metadata rule correctly 1`] = `"Address marked with "test=test""`;

exports[`request-policies.composable > display the allowlisted rule correctly 1`] = `"Address in address book"`;

exports[`request-policies.composable > display the any user percentage quorum rule correctly 1`] = `"50% approval of all users"`;

exports[`request-policies.composable > display the any user quorum rule correctly 1`] = `"5 approvals from any user"`;

exports[`request-policies.composable > display the auto approved rule correctly 1`] = `"No approval required"`;

exports[`request-policies.composable > display the some groups quorum rule correctly 1`] = `"1 approval from Admin, Developers"`;

exports[`request-policies.composable > display the some users percentage quorum rule correctly 1`] = `"50% approval from Alice"`;

exports[`request-policies.composable > display the some users quorum rule correctly 1`] = `"5 approvals from Alice, Bob"`;
